server/lib/scheduler/Scheduler.ts

```diff
@@ -0,0 +1,110 @@
+import { filter } from "asyncro";
+
+class Scheduler {
+  private fetchAllWorkspaces;
+  private rootParentCache;
+  private schedule;
+
+  public constructor({ fetchAllWorkspaces, rootParentCache, schedule }) {
+    this.fetchAllWorkspaces = fetchAllWorkspaces;
+    this.rootParentCache = rootParentCache;
+    this.schedule = schedule;
+  }
+
+  public async getIdOfCurrentWorkspace(userId) {
+    const assignment = this.schedule.getMostRecentAssignmentForUser(userId);
+    return assignment.getWorkspaceId();
+  }
+
+  public async findNextWorkspace(userId) {
+    // clear cache so we don't use old eligibility info
+    this.rootParentCache.clearRootParentCache();
+    const idsOfWorkspacesThatCouldBeNext = await this.getIdsOfWorkspacesThatCouldBeNext(userId);
+    const assignedWorkspaceId = this.pickWorkspaceIdAtRandom(idsOfWorkspacesThatCouldBeNext);
+    await this.schedule.assignWorkspaceToUser(userId, assignedWorkspaceId);
+  }
+
+  private async getIdsOfWorkspacesThatCouldBeNext(userId) {
+    const allWorkspaces = await this.fetchAllWorkspaces();
+    const allEligibleWorkspaces = await this.filterByEligibility(allWorkspaces);
+    const workspacesInTreeWorkedOnLeastRecently = await this.filterByWhetherInTreeWorkedOnLeastRecently(allEligibleWorkspaces);
+    const notYetWorkedOnInThatTree = await this.filterByWhetherNotYetWorkedOn(workspacesInTreeWorkedOnLeastRecently);
+
+    let finalWorkspaces = notYetWorkedOnInThatTree;
+
+    // if every workspace in that tree has been worked on
+    // then instead look for workspaces with remaining budgets
+    if (finalWorkspaces.length === 0) {
+      finalWorkspaces = await this.filterByWhetherHasRemainingBudget(workspacesInTreeWorkedOnLeastRecently);
+    }
+
+    // if every workspace in that tree has 0 remaining budget
+    // then instead look for the workspace worked on least recently
+    if (finalWorkspaces.length === 0) {
+      const idOfWorkspaceWorkedOnLeastRecently = await this.schedule.getLeastRecentlyActiveWorkspace(workspacesInTreeWorkedOnLeastRecently.map(w => w.id));
+      return [idOfWorkspaceWorkedOnLeastRecently];
+    }
+
+    return finalWorkspaces.map(w => w.id);
+  }
+
+  private pickWorkspaceIdAtRandom(workspaceIds) {
+    const randomIndex = Math.floor(Math.random() * (workspaceIds.length));
+    return workspaceIds[randomIndex];
+  }
```

Comment:
This screams "utility method" to me; you could rename it ```pickRandomItem``` and make it generic for all arrays, and place it above. In part, to make it apparent that what it's doing is generic.